crates/fuel-core/src/schema/tx/types.rs

```diff
@@ -620,6 +624,71 @@ impl Transaction {
     }
 }
 
+#[derive(Union, Debug)]
+pub enum DryRunTransactionStatus {
+    Success(DryRunSuccessStatus),
+    Failed(DryRunFailureStatus),
+}
+
+impl DryRunTransactionStatus {
+    pub fn new(tx_status: TransactionExecutionResult) -> Self {
+        match tx_status {
+            TransactionExecutionResult::Success { result } => {
+                DryRunTransactionStatus::Success(DryRunSuccessStatus { result })
+            }
+            TransactionExecutionResult::Failed { result, reason } => {
+                DryRunTransactionStatus::Failed(DryRunFailureStatus { result, reason })
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct DryRunSuccessStatus {
+    result: Option<VmProgramState>,
+}
+
+#[Object]
+impl DryRunSuccessStatus {
+    async fn program_state(&self) -> Option<ProgramState> {
+        self.result.map(Into::into)
+    }
+}
+
+#[derive(Debug)]
+pub struct DryRunFailureStatus {
+    result: Option<VmProgramState>,
+    reason: String,
+}
+
+#[Object]
+impl DryRunFailureStatus {
+    async fn program_state(&self) -> Option<ProgramState> {
+        self.result.map(Into::into)
+    }
+
+    async fn reason(&self) -> String {
+        self.reason.clone()
+    }
+}
+
+pub struct DryRunTransaction(pub TransactionExecutionStatus);
+
+#[Object]
+impl DryRunTransaction {
+    async fn id(&self) -> TransactionId {
+        TransactionId(self.0.id)
+    }
+
+    async fn status(&self) -> async_graphql::Result<DryRunTransactionStatus> {
```

Comment:
I don't think these two need to be fallible. I think they are just doing a conversion.